go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+	From  sqlparser.TableExpr
+	order int
+	// Route is the plan object being built. It will contain all the
+	// information necessary to execute the route operation.
+	Route *Route
+}
+
+// Order returns the order of the node.
+func (rtb *RouteBuilder) Order() int {
+	return rtb.order
+}
+
+// MarshalJSON marshals RouteBuilder into a readable form.
+// It's used for testing and diagnostics. The representation
+// cannot be used to reconstruct a RouteBuilder.
+func (rtb *RouteBuilder) MarshalJSON() ([]byte, error) {
+	marshalRoute := struct {
+		From  string `json:",omitempty"`
+		Order int
+		Route *Route
+	}{
+		From:  sqlparser.String(rtb.From),
+		Order: rtb.order,
+		Route: rtb.Route,
+	}
+	return json.Marshal(marshalRoute)
+}
+
+// Route is a Plan object that represents a route.
+// It can be any one of the Select primitives from PlanID.
+// Some plan ids correspond to a multi-shard query,
+// and some are for a single-shard query. The rules
+// of what can be merged, or what can be pushed down
+// depend on the PlanID. They're explained in code
+// where such decisions are made.
+// TODO(sougou): struct is incomplete.
+// TODO(sougou): integrate with the older v3 Plan.
+type Route struct {
+	// PlanID will be one of the Select IDs from PlanID.
+	PlanID PlanID
+	// Keypsace represents the keyspace to which
+	// the query will be sent.
+	Keyspace *Keyspace
+	// Vindex represents the vindex that will be used
+	// to resolve the route.
+	Vindex Vindex `json:",omitempty"`
+	// Values represents a single value or a list of
+	// values that will be used as input to the Vindex
+	// to compute the target shard(s) where the query must
+	// be sent.
+	// TODO(sougou): explain contents of Values.
+	Values interface{} `json:",omitempty"`
+}
+
+// buildSelectPlan2 is the new function to build a Select plan.
+// TODO(sougou): rename after deprecating old one.
+func buildSelectPlan2(sel *sqlparser.Select, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	return processTableExprs(sel.From, schema)
+}
+
+// processTableExprs analyzes the FROM clause. It produces a PlanBuilder
+// and the associated SymbolTable with all the routes identified.
+func processTableExprs(tableExprs sqlparser.TableExprs, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	if len(tableExprs) != 1 {
+		// TODO(sougou): better error message.
+		return nil, nil, errors.New("no list")
+	}
+	return processTableExpr(tableExprs[0], schema)
+}
+
+// processTableExpr produces a PlanBuilder subtree and SymbolTable
+// for the given TableExpr.
+func processTableExpr(tableExpr sqlparser.TableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch tableExpr := tableExpr.(type) {
+	case *sqlparser.AliasedTableExpr:
+		return processAliasedTable(tableExpr, schema)
+	case *sqlparser.ParenTableExpr:
+		planBuilder, symbols, err := processTableExprs(tableExpr.Exprs, schema)
+		if route, ok := planBuilder.(*RouteBuilder); ok {
```

Comment:
I see you’re setting the “From” of the route child to point to the Parenthesized table expr. Can you explain why in a comment?